sdk/examples/xo_javascript/xo_handler.js

```diff
@@ -0,0 +1,271 @@
+  _display(`Player ${player.toString().substring(0, 6)} created a game.`)
+  return _setEntry(state, address, setValue)
+}
+
+const _handleTake = (state, address, update, player) => (possibleAddressValues) => {
+  let stateValueRep = possibleAddressValues[address]
+  let stateValue
+  stateValue = _decodeState(stateValueRep)
+  if (stateValue.Board == "") {
+    throw new InvalidTransaction("Invalid Action: Take requires an existing game.")
+  }
+  if (['P1-WIN', 'P2-WIN', 'TIE'].includes(stateValue.GameState)){
+    throw new InvalidTransaction("Invalid Action: Game has ended.")
+  }
+
+  if (!(['P1-NEXT', 'P2-NEXT'].includes(stateValue.GameState))) {
+    throw new InternalError(`Game has reached an invalid state: ${stateValue.GameState}`)
+  }
+
+  if (stateValue.Player1 === "") {
+    stateValue.Player1 = player
+  }
+
+  else if (stateValue.Player2 === "") {
+      stateValue.Player2 = player
+  }
+  let boardList = stateValue.Board.split("")
+
+  if (boardList[update.Space - 1] != "-") {
+    throw new InvalidTransaction("Invalid Action: Space already taken.")
+  }
+
+  if (stateValue.GameState === "P1-NEXT" && player == stateValue.Player1) {
+    boardList[update.Space - 1] = "X"
+    stateValue.GameState = "P2-NEXT"
+  }
+  else if (stateValue.GameState === "P2-NEXT" && player == stateValue.Player2) {
+    boardList[update.Space - 1] = "O"
+    stateValue.GameState = "P1-NEXT"
+  }
+  else {
+    throw new InvalidTransaction(`Not this player's turn: ${player.toString().substring(0, 6)}`)
+  }
+
+  stateValue.Board = boardList.join("")
+
+  if (_isWin(stateValue.Board, "X")) {
+    stateValue.GameState = "P1-WIN"
+  }
+  else if (_isWin(stateValue.Board, "O")) {
+    stateValue.GameState = "P2-WIN"
+  }
+  else if (stateValue.Board.search('-') == -1) {
+    stateValue.GameState = "TIE"
+  }
+
+  let setValue = Buffer.from([stateValue.Board,stateValue.GameState, stateValue.Player1,
+      stateValue.Player2, update.Name].join())
+  _display(`Player ${player.toString().substring(0, 6)} takes space: ${update.Space}\n\n` +
+           _gameToStr(stateValue.Board, stateValue.GameState, stateValue.Player1
+             , stateValue.Player2, update.Name))
+
+  return _setEntry(state, address, setValue)
+}
+
+class XOHandler extends TransactionHandler {
+  constructor () {
+    super(XO_FAMILY, '1.0', 'csv-utf8', [XO_NAMESPACE])
+  }
+
+  apply (transactionProcessRequest, state) {
+    return _decodeRequest(transactionProcessRequest.payload)
+      .catch(_toInternalError)
+      .then((update) => {
+        let header = TransactionHeader.decode(transactionProcessRequest.header)
+        let player = header.signerPubkey
+        if (!update.Name) {
+          throw new InvalidTransaction('Name is required')
+        }
+
+        if (!update.Action) {
+          throw new InvalidTransaction('Action is required')
+        }
+
+        else if (update.Action === "take") {
+          try{
```

Comment:
This code should probably be part of the `_handleTake` function, as it seems a big confusing here.